game/src/info/trip.rs

```diff
@@ -386,6 +389,329 @@ pub fn cancelled(ctx: &mut EventCtx, app: &App, id: TripID) -> Widget {
+        msgs.push(format!("  Started at {}", p.start_time.ampm_tostring()));
+        let phase_duration = if let Some(t2) = p.end_time {
+            let d = t2 - p.start_time;
+            msgs.push(format!(
+                "  Ended at {} (duration: {})",
+                t2.ampm_tostring(),
+                d
+            ));
+            d
+        } else {
+            let d = app.primary.sim.time() - p.start_time;
+            msgs.push(format!("  Ongoing (duration so far: {})", d));
+            d
+        };
+        // TODO Problems when this is really low?
+        let percent_duration = if total_duration_so_far == Duration::ZERO {
+            0.0
+        } else {
+            phase_duration / total_duration_so_far
+        };
+
+        let phase_width = box_width * percent_duration;
+        if idx == phases.len() - 1 {
+            if let Some(p) = progress_along_path {
+                icons.push(Widget::draw_batch(
+                    ctx,
+                    GeomBatch::load_svg(ctx.prerender, "system/assets/timeline/current_pos.svg")
+                        .centered_on(Pt2D::new(p * phase_width, 7.5)),
+                ));
+            }
+        }
+        icons_geom.push(GeomBatch::load_svg(
+            ctx.prerender,
+            match p.phase_type {
+                TripPhaseType::Driving => "system/assets/timeline/driving.svg",
+                TripPhaseType::Walking => "system/assets/timeline/walking.svg",
+                TripPhaseType::Biking => "system/assets/timeline/biking.svg",
+                TripPhaseType::Parking => "system/assets/timeline/parking.svg",
+                TripPhaseType::WaitingForBus(_, _) => "system/assets/timeline/waiting_for_bus.svg",
+                TripPhaseType::RidingBus(_, _, _) => "system/assets/timeline/riding_bus.svg",
+                TripPhaseType::Cancelled | TripPhaseType::Finished => unreachable!(),
+                TripPhaseType::DelayedStart => "system/assets/timeline/delayed_start.svg",
+                // TODO What icon should represent this?
+                TripPhaseType::Remote => "system/assets/timeline/delayed_start.svg",
+            },
+        ));
+        msgs.push(format!(
+            "  {}% of trip percentage",
+            (100.0 * percent_duration) as usize
+        ));
+
+        msgs.push(format!(
+            "  Total delayed time {}",
+            app.primary.sim.trip_blocked_time(trip_id)
+        ));
+        let mut sum_phase_dist = Distance::ZERO;
+        if let Some((_, step_list)) = &p.path {
+            let mut norm_distance = Distance::meters(0.0);
+            for step in step_list.get_steps() {
+                match step {
+                    PathStep::Lane(id) | PathStep::ContraflowLane(id) => {
+                        let lane_detail = map.get_l(*id);
+                        sum_phase_dist += lane_detail.length();
+                        let mut found = false;
+                        for (lane, avg_speed) in &data_lane {
```

Comment:
Whoa, this might be slow. For every step in the path, we're checking all of the delayed lanes. Since the order of the delayed lanes isn't interesting, what if the method in `Analytics` returned a map from lane ID?

Reply:
A map would be so much better!

Reply:
It's an awkward lifetime problem I haven't figured out. I think the clearest way is
```
let blank_speeds = BTreeMap::new();
let lane_speeds =  app.primary.sim.get_analytics().trip_lane_speeds.get(&trip_id).unwrap_or(&blank_speeds);
```
It's not expensive to make a blank btreemap on the stack, especially since this code isn't called in a tight inner loop or anything

Reply:
Yup, has been changed

Reply:
Thanks